### PR TITLE
Improve field validation for documents

### LINKED ISSUE #720: bson.errors.invalidDocument: key * must not contains '.'
When using a real mongodb database I can include a entry like a below, but in mongomock it not possible.

{"object.name": value} 

When I add this in real database it works, but in mongomock doesn`t.

### PR DESCRIPTION
**Pull request by plaan**
Starting from mongodb 3.6 field names cannot contain the null character and top-field names cannot start with the dollar sign. Apart form those cases mongodb does permit storage of the dots and dollar signs ([mongodb v3.6](https://www.mongodb.com/docs/v3.6/reference/limits/#naming-restrictions)). Newer versions of mongodb add improved support ([mongodb v4.0](https://www.mongodb.com/docs/v4.0/reference/limits/#naming-restrictions), [mongodb v5.0](https://www.mongodb.com/docs/manual/reference/limits/)).

Fixes: https://github.com/mongomock/mongomock/issues/720

### CODE REVIEW

File: mongomock/collection.py
Comment (pcorpet) [inline mongomock/collection.py]: Use `key.startswith`

File: mongomock/collection.py
Comment (pcorpet) [inline mongomock/collection.py]: This function isn't using any closure, so I'd rather not have it as a nested function

File: mongomock/collection.py
Comment (pcorpet) [inline mongomock/collection.py]: Please use single quotes for literal strings. `'\0'` (same in all your changes).

**Review by pcorpet — CHANGES_REQUESTED** [review]
Thanks a lot for your contribution.

Please add a test that illustrates that recent versions accept weirder names.

File: mongomock/collection.py
Comment (plaan) [reply mongomock/collection.py]: Forgot about that one, thanks!

File: mongomock/collection.py
Comment (plaan) [reply mongomock/collection.py]: Moved it out of the function

File: mongomock/collection.py
Comment (plaan) [reply mongomock/collection.py]: Changed to using all single quotes

**Comment by plaan** [comment]
Hi @pcorpet, wondering if you have the time for a second review. I made sure to add your suggested changes!

**Comment by lluissalord** [comment]
Hi @pcorpet @plaan do we have any update on this? We are encountering with this issue, and we would like to solve it.

**Comment by josemarcosrf** [comment]
This would be a great one so the mocking gets closer to the validation of the actual mongo. Any updates on getting this one merged?

**Comment by pcorpet** [comment]
Thanks for pinging folks, it's working. I'll take a look now.

**Review by pcorpet — APPROVED** [review]

**Pull request merged**
